lib/sparse/src/index/search_context.rs

```diff
@@ -0,0 +1,454 @@
+                    score +=
+                        element.weight * self.query.weights[posting_iterator.query_weight_offset];
+                }
+            }
+        }
+
+        Some(ScoredCandidate {
+            score,
+            vector_id: min_record_id,
+        })
+    }
+
+    pub fn next_min(to_inspect: &[IndexedPostingListIterator<'_>]) -> Option<u32> {
+        // Initialize min record id with max value
+        let mut min_record_id = u32::MAX;
+        // Indicates that posting iterators are not empty
+        let mut found = false;
+
+        // Iterate first time to find min record id at the head of the posting lists
+        for posting_iterator in to_inspect.iter() {
+            if let Some(element) = posting_iterator.posting_list_iterator.peek() {
+                found = true;
+                if element.id < min_record_id {
+                    min_record_id = element.id;
+                }
+            }
+        }
+
+        if !found {
+            return None;
+        }
+        Some(min_record_id)
+    }
+
+    /// Make sure the longest posting list is at the head of the posting list iterators
+    pub fn sort_posting_lists_by_len(&mut self) {
+        // decreasing order
+        self.postings_iterators.sort_by(|a, b| {
+            b.posting_list_iterator
+                .len_left()
+                .cmp(&a.posting_list_iterator.len_left())
+        });
+    }
+
+    pub fn search(&mut self) -> Vec<ScoredCandidate> {
+        while let Some(candidate) = self.advance() {
+            // push candidate to result queue
+            self.result_queue.push(candidate);
+
+            // we potentially have enough results to prune low performing posting lists
+            if self.result_queue.len() == self.top {
+                // current min score
+                let min_score = self.result_queue.top().unwrap().score;
+
+                // sort posting lists by length to try to prune the longest one
+                self.sort_posting_lists_by_len();
+
+                self.prune_longest_posting_list(min_score);
+            }
+        }
+        // posting iterators exhausted, return result queue
+        let queue = std::mem::take(&mut self.result_queue);
+        queue.into_vec()
+    }
+
+    /// Prune posting lists that cannot possibly contribute to the top results
+    /// Assumes longest posting list is at the head of the posting list iterators
+    /// Returns true if the longest posting list was pruned
+    pub fn prune_longest_posting_list(&mut self, min_score: f32) -> bool {
+        // compute skip target before acquiring mutable reference to posting list iterator
+        let skip_to = if self.postings_iterators.len() == 1 {
+            // if there is only one posting list iterator, we can skip to the end
+            None
+        } else {
+            // otherwise, we skip to the next min elements in the remaining posting list
+            Self::next_min(&self.postings_iterators[1..])
+        };
```

Comment:
Could be simplified to this, but not sure if you like it.

```suggestion
        // compute skip target before acquiring mutable reference to posting list iterator
        let skip_to = (self.postings_iterators.len() != 1)
            .then(|| Self::next_min(&self.postings_iterators[1..]));
```

Reply:
Thanks for the proposal, 
TBH I am not a fan of the bool to Option combinators but I have noticed that you like those :laughing:

Reply:
> TBH I am not a fan of the bool to Option combinators but I have noticed that you like those laughing

I like things short. For me that makes it easier to glance what's going on. ^^

Reply:
I'm with Arnaud on this one. These bool methods are confusing af. 🙈